mailgun-relayery/src/servers/channels/database.go

```diff
@@ -0,0 +1,310 @@
+package channels
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/bmatsuo/lmdb-go/lmdb"
+	"github.com/bmatsuo/lmdb-go/lmdbscan"
+	"github.com/golang/protobuf/proto"
+
+	"servers/protoed_channel"
+)
+
+const dbChannelName = "channelDb"
+
+// Key holds a key at which data is stored in the database.
+type Key struct {
+	descriptor string
+}
+
+// Encode encodes the key to an array of bytes.
+func (k *Key) Encode() ([]byte, error) {
+	bb := make([]byte, 0, len(k.descriptor))
+
+	w := bytes.NewBuffer(bb)
+	_, err := w.WriteString(k.descriptor)
+	if err != nil {
+		return nil, err
+	}
+
+	bb = w.Bytes()
+
+	return bb, nil
+}
+
+// DecodeKey decodes an array of bytes to a key.
+func DecodeKey(data []byte) (key Key, err error) {
+	key.descriptor = string(data)
+	return
+}
+
+// newEnv initializes an lmdb database environment.
+func newEnv() (env *lmdb.Env, err error) {
+	env, err = lmdb.NewEnv()
+	if err != nil {
+		err = fmt.Errorf("failed to initialize a database environment: %s", err)
+		return
+	}
+
+	err = env.SetMaxDBs(1)
+	if err != nil {
+		err = fmt.Errorf("failed to set the max. number of DBs to 1: %s", err)
+		closeErr := env.Close()
+		if closeErr != nil {
+			err = fmt.Errorf("%s; failed to close the database: %s", err, closeErr)
+		}
+		return
+	}
+
+	mapSize := int64(32 * 1024 * 1024 * 1024)
+	err = env.SetMapSize(mapSize)
+	if err != nil {
+		err = fmt.Errorf("failed to set the map size to %d: %s", mapSize, err)
+		closeErr := env.Close()
+		if closeErr != nil {
+			err = fmt.Errorf("%s; failed to close the database: %s", err, closeErr)
+		}
+		return
+	}
+
+	return
+}
+
+// Database represents a database of channels.
+type Database struct {
+	Path     string
+	env      *lmdb.Env
+	Readonly bool
+}
+
+// Initialize initializes the database environment in the given directory.
+// Creates the expected database and should be called only once during the deployment.
+func Initialize(path string) (err error) {
+	_, err = os.Stat(path)
+	if err != nil {
+		return
+	}
+
+	env, err := newEnv()
+	if err != nil {
+		return
+	}
+	defer func() {
+		err = env.Close()
+	}()
+
+	err = env.Open(path, 0, 0660)
+	if err != nil {
+		return
+	}
+
+	err = env.Update(func(txn *lmdb.Txn) (txnErr error) {
+		_, txnErr = txn.OpenDBI(dbChannelName, lmdb.Create)
+		if txnErr != nil {
+			return
+		}
+		return
+	})
+
+	return
+}
+
+// NewDatabase creates a new database object. The database directory is assumed to be already initialized.
+func NewDatabase(
+	path string,
+	readonly bool) (d *Database, err error) {
+
+	env, err := newEnv()
```

Comment:
What would be the advantages to this approach? It seems counterintuitive to me not to have a `Database` struct directly.

Reply:
(After our discussion):

I'm still confused about the nature of this change - I think you meant to say that we won't need NewDatabase anyomre, but Initialize stays in place?

Reply:
Since there is no _database_ specific logic (as a database in LMDB context -- a directory is tied to an LMDB _environment_, not an LMDB _database_), you don't need `Database` struct nor `NewDatabase`. What you do need, on the other hand, is `NewEnv(.)`, struct `Env` and `InitializeEnv(.)`.